libazureinit/src/provision/mod.rs

```diff
@@ -23,156 +24,116 @@ use crate::User;
-    /// order provided until one succeeds.
-    pub fn hostname_provisioners(
-        mut self,
-        backends: impl Into<Vec<hostname::Provisioner>>,
-    ) -> Self {
-        self.hostname_backends = Some(backends.into());
-        self
-    }
-
-    /// Specify the ways to create a user in the virtual machine
-    ///
-    /// By default, all known methods will be attempted. Use this function to
-    /// restrict which methods are attempted. These will be attempted in the
-    /// order provided until one succeeds.
-    pub fn user_provisioners(
-        mut self,
-        backends: impl Into<Vec<user::Provisioner>>,
-    ) -> Self {
-        self.user_backends = Some(backends.into());
-        self
-    }
-
-    /// Specify the ways to set a users password.
-    ///
-    /// By default, all known methods will be attempted. Use this function to
-    /// restrict which methods are attempted. These will be attempted in the
-    /// order provided until one succeeds. Only relevant if a password has been
-    /// provided with the [`User`].
-    pub fn password_provisioners(
-        mut self,
-        backend: impl Into<Vec<password::Provisioner>>,
-    ) -> Self {
-        self.password_backends = Some(backend.into());
-        self
-    }
-
-    /// Provision the host.
-    ///
-    /// Provisioning can fail if the host lacks the necessary tools. For example,
-    /// if there is no `useradd` command on the system's `PATH`, or if the command
-    /// returns an error, this will return an error. It does not attempt to undo
-    /// partial provisioning.
     #[instrument(skip_all)]
     pub fn provision(self) -> Result<(), Error> {
-        self.user_backends
-            .unwrap_or_else(|| user::Provisioner::iter().collect())
+        self.config
+            .hostname_provisioners
+            .backends
             .iter()
-            .find_map(|backend| {
-                backend
-                    .create(&self.user)
-                    .map_err(|e| {
-                        tracing::info!(
-                            error=?e,
-                            backend=?backend,
-                            resource="user",
-                            "Provisioning did not succeed"
-                        );
-                        e
-                    })
-                    .ok()
+            .find_map(|backend| match backend {
+                HostnameProvisioner::Hostnamectl => {
+                    hostname::Provisioner::Hostnamectl.set(&self.hostname).ok()
+                }
+                #[cfg(test)]
+                HostnameProvisioner::FakeHostnamectl => Some(()),
+            })
+            .ok_or(Error::NoHostnameProvisioner)?;
+
+        self.config
+            .user_provisioners
+            .backends
+            .iter()
+            .find_map(|backend| match backend {
+                UserProvisioner::Useradd => {
+                    user::Provisioner::Useradd.create(&self.user).ok()
```

Comment:
Do we want to panic here if the user cannot be created?

Reply:
I would say the same here as I did above -- was trying to mirror the original behavior and maintain predictability in behavior.  But, if people think panicking might make the error more immediately obvious, then can panic instead of returning ok_or(Error::NoUserProvisioner)!